src/search.coffee

```diff
@@ -6,234 +6,257 @@ url = require 'url'
+          return
+        result.id = matches[1]
+      return callback 'Unable to parse an ID out of result URL.' if invalidId
+
+      # Return search count and results.
+      callback null, {
+        count: body.totalEstimatedMatches,
+        results: body.value
+      }
 
     debug url.format req.uri
 
-  counts: (query, callback) ->
-    getCounts = (options, callback) =>
-      options = _.extend {}, options, {query, sources: Search.SOURCES}
-      @search 'Composite', options, (err, result) =>
-        return callback err if err
-        callback null, @extractCounts result
-
-    # Two requests are needed. The first request is to get an accurate
-    # web results count and the second request is to get an accurate count
-    # for the rest of the verticals.
-    #
-    # The 1,000 value comes from empirical data. It seems that after 600
-    # results, the accuracy gets quite consistent and accurate. I picked 1,000
-    # just to be in the clear. It also doesn't matter if there are fewer than
-    # 1,000 results.
-    async.map [{skip: 1000}, {}], getCounts, (err, results) ->
-      return callback err if err
-      callback null, _.extend results[1], _.pick(results[0], 'web')
-
-  extractCounts: (result) ->
-    keyRe = /(\w+)Total$/
-
-    _.chain(result?.d?.results or [])
-      .first()
-      .pairs()
-      .filter ([key, value]) ->
-        # Eg. WebTotal, ImageTotal, ...
-        keyRe.test key
-      .map ([key, value]) ->
-        # Eg. WebTotal => web
-        key = keyRe.exec(key)[1].toLowerCase()
-        value = Number value
-
-        switch key
-          when 'spellingsuggestions' then ['spelling', value]
-          else [key, value]
-      .object()
-      .value()
-
-  verticalSearch: (vertical, verticalResultParser, query, options, callback) ->
-    [callback, options] = [options, {}] if _.compact(arguments).length is 4
-
-    @parallelSearch vertical, _.extend({}, options, {query}), (err, result) ->
-      return callback err if err
-      callback null, verticalResultParser result
-
-  mapResults: (results, fn) ->
-    _.chain(results)
-      .pluck('d')
-      .pluck('results')
-      .flatten()
-      .map fn
-      .value()
+  _parallelSearch: (options, callback) ->
+    allRequestOptions = []
+
+    top = Search.MAX_RESULTS
+    if options.top?
+      top = Number options.top
+      options = _.omit options, 'top'
+
+    allRequestOptions.push _.defaults {
+      count: Math.min Search.PAGE_SIZE, top - offset
+      offset
+    }, options for offset in [(options.offset || 0)...top] by Search.PAGE_SIZE
+
+    async.mapLimit allRequestOptions, @parallelLimit, _.bind(@_executeSearch, this),
+      (err, responses) ->
+        callback err if err?
+
+        data = {
+          count: 0
+          results: []
+        }
+
+        # Avoid duplicates by checking result IDs.
+        existingIds = []
+        responses.forEach (response) ->
+          data.count = response.count if response.count?
```

Comment:
I'm not saying there isn't a better, or at least more obvious, way to do this...but, here is what is happening:
 - `_parallelSearch()` handles the creation of asynchronous calls to `_executeSearch()` whenever the user-facing methods, `counts()`/`web()`/`images()`/`videos()`/`news()`, are called.
 - Within this callback, we do some basic sanitation of `body.value` (the resultset) by removing any duplicates.
 - Then, we return the list of results along with the `totalEstimatedMatches` (which is used by `counts()`).

I chose to reset/overwrite the count each time because there's a chance that the accuracy of `totalEstimatedMatches` improves as you request later pages (i.e. a higher `offset`). Realistically, I think this just needs to be made more obvious. That can probably be done by taking that line out of the `forEach` like so:

```coffeescript
data = 
  count: _.last(responses).count
  results: [] 
```

Also, I'll go ahead and update `count` to a more appropriate `estimatedCount` since it is not necessarily equal/related to `results.length`.